### PR TITLE
added secure deployment guide

### LINKED ISSUE #185: Document the security best practice
We need to document the security best practice to tell users how to use Notation in a security posture. The security best practice and considerations include but are not limited to:

- Security considerations for using Notation plugins
- Security considerations for using container registries
- Notation authentication (Credential Store)
- What should users do against the threat model https://github.com/notaryproject/notaryproject/pull/242

This is a big topic. We can iterate on the content creation.

### LINKED ISSUE #224: Document security best practices for Notation deployment
As part of the release of Notation 1.0.0 we need to document the security best practices to use Notation in various scenarios. One important part is to document how to protect the locations on the file system where the policies, plugins, trust store, etc. are located and what are the associated risks of unauthorized access to those locations.

### PR DESCRIPTION
**Pull request by zr-msft**
fixes #224 
fixes #185

### CODE REVIEW

File: content/en/docs/concepts/secure-deployment.md
Comment (yizha1) [inline content/en/docs/concepts/secure-deployment.md:21]: nit:
`in on the filesystem` ==> `in the filesystem`

**Review by yizha1 — COMMENTED** [review]
Thanks @zr-msft

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: nice catch, fixed

Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: reworked the text to include "general best practices" language

File: content/en/docs/concepts/secure-deployment.md
Comment (toddysm) [inline content/en/docs/concepts/secure-deployment.md:8]: Shouldn't this be:
> ... or in a CI/CD pipeline...

File: content/en/docs/concepts/secure-deployment.md
Comment (toddysm) [inline content/en/docs/concepts/secure-deployment.md:12]: Maybe we should link to the documentation for those

File: content/en/docs/concepts/secure-deployment.md
Comment (toddysm) [inline content/en/docs/concepts/secure-deployment.md:13]: Aren't certs in the trust stores? Why do we have separate bullet for certificates?

File: content/en/docs/concepts/secure-deployment.md
Comment (toddysm) [inline content/en/docs/concepts/secure-deployment.md:19]: I think we should get more into details like:
- Only Notation user having access to those folders (or sudo)

Also, we need to explain what are the risks of somebody accessing any of those folders. Like replacing plugin with malicious one, changing the policy to allow images signed with malicious key, adding untrusted certs etc.

**Review by toddysm — CHANGES_REQUESTED** [review]

**Review by shizhMSFT — DISMISSED** [review]
LGTM but I'm not a maintainer.

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md:33]: @toddysm  can you elaborate or point me towards where those are defined?

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: fixed

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: fixed

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: fixed

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md:18]: fixed

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: fixed

File: content/en/docs/concepts/secure-deployment.md
Comment (byronchien) [inline content/en/docs/concepts/secure-deployment.md:26]: nit: empty bullet point should be removed

File: content/en/docs/concepts/secure-deployment.md
Comment (byronchien) [inline content/en/docs/concepts/secure-deployment.md:16]: i think that the risks of not following secure best practices can be called out as part of this initial section. example:

> You are responsible for securing your deployment of Notation and following the best practices outlined in this document. Failure to do so can result in malicious plugins being installed, trust policies being modified to allow verification compromised or untrusted artifacts, trust stores being modified to have untrusted or compromised certificates, or signing keys being compromised.

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: nice catch, fixed

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: Good call, relocated content to this section

File: content/en/docs/concepts/secure-deployment.md
Comment (byronchien) [inline content/en/docs/concepts/secure-deployment.md:14]: will this link to [this page](https://github.com/notaryproject/notaryproject.dev/pull/219)?

**Review by byronchien — APPROVED** [review]

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md:14]: @byronchien yes, as soon as the PR is merged. I'll provide the update in a follow up PR.

File: content/en/docs/concepts/secure-deployment.md
Comment (priteshbandi) [inline content/en/docs/concepts/secure-deployment.md:16]: Should we callout Notation works in shared responsibility model. Users are responsible for securing the host on which notation is deployed ?

Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md:16]: @priteshbandi  i feel like this line states this fact without calling it "shared responsibility". Do you feel like we need to state this more explicitly?

Comment (priteshbandi) [reply content/en/docs/concepts/secure-deployment.md:16]: Yeah we should be explicit about shared responsibility.

Comment (yizha1) [reply content/en/docs/concepts/secure-deployment.md:16]: @zr-msft Please check this comment from Pritesh on threat modeling PR, https://github.com/notaryproject/notaryproject/pull/242#discussion_r1223622951, it will help you to understand shared responsibility.

File: content/en/docs/concepts/secure-deployment.md
Comment (toddysm) [reply content/en/docs/concepts/secure-deployment.md:35]: We do not have those defined :) In the above text we say: "use your organization best practices" but we don't say what our community's recommended best practices are.

Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md:35]: @toddysm i'd like to have this PR merged for now and we can do a follow up PR once the community best practices are defined. Does that work for you?

**Review by yizha1 — DISMISSED** [review]
LGTM

**Comment by yizha1** [comment]
@priteshbandi @iamsamirzon could you review this PR again?

**Review by toddysm — COMMENTED** [review]
The purpose of this document is to provide prescriptive guidance for the deployment. I am not sure we are achieving that goal with the current text and am not sure what the value of this documentation is. Here are the main questions that I think should be answered:
- How should I obtain the binaries (I believe we are covering that in the installation guide but reinforcing this here and making sure that the the chksum is validated is imperative here)
- If there is no installation, how do I copy the binary to the respective folder? Do I use sudo or not? Should I create new user or not? Should I change the folder permissions or not?
- How do I set up the signing key? What user do I use for that? Should I use sudo or separate user?
- For validation - how do I download the cert? What user do I use to download the cert? Should I protect the folder where the cert is stored? What permissions should the folder have?
- For trust policies - what user do I use to create the policy? Should I use sudo or not? What permissions the folder where the policy is should have?

**Comment by shizhMSFT** [comment]
All above questions are PM questions but I can provide some security suggestions:

> How should I obtain the binaries (I believe we are covering that in the installation guide but reinforcing this here and making sure that the the chksum is validated is imperative here)

In general, binaries can be obtained from any sources, either trusted or untrusted. It is critical to obtain the checksum from a trusted source so that the binaries can be verified. Overall, the binaries can be obtained from a package manager with authenticity guarantee (e.g. `brew`) or downloaded directly from github.com via HTTPS.

> If there is no installation, how do I copy the binary to the respective folder? Do I use sudo or not? Should I create new user or not? Should I change the folder permissions or not?

It depends on whether you want to install for all users on the system or just you. If it is for all users on the system, it is better to copy the binary to `/usr/local/bin` with the owner set to `root` and the permission set to `rwxr-xr-x` (`755`). `sudo` is required if you are not `root`. If it is just for you, you can copy it to anywhere with the owner remaining set to yourself and permission set to at least `rwx------` (`700`).

> How do I set up the signing key? What user do I use for that? Should I use sudo or separate user?

Since you are setting up the signing key for yourself, you should use your own account to do that. Never use `sudo` or as a separate user unless you are an admin and try to set up the signing key for that user. This suggestion applies regardless whether the binary is installed for all users on the system or just for you.

> For validation - how do I download the cert? What user do I use to download the cert? Should I protect the folder where the cert is stored? What permissions should the folder have?

The certs must be downloaded from trusted sources. It can be done by directly downloading from the web page via HTTPS from the signing key vendor or by running scripts or tools provided by the signing key vendor. Since those certs may be private certs (e.g. no CT records) and contain confidential information, it is recommended to set the permissions to `rw-------` (`600`) to prevent access from other users.

> For trust policies - what user do I use to create the policy? Should I use sudo or not? What permissions the folder where the policy is should have?

It depends on the precise scenarios whether the trust policies are for all users on the system or just for you. The first scenario is not defined yet so that I would like to skip for now. For the second scenario, you should create the policy using your own account without `sudo`. For privacy consideration, you should keep the folder with permission set to `rwx------` (`700`) and the policy file with permission set to `rw-------` (`600`)

**Comment by yizha1** [comment]
Thanks @shizhMSFT for providing the comments. I recalled one [issue](https://github.com/notaryproject/notation/issues/203), that only user level config is supported. The system level config is not supported and there is no plan for now. So, we just need to document user level config as you explained.

File: content/en/docs/concepts/secure-deployment.md
Comment (yizha1) [reply content/en/docs/concepts/secure-deployment.md:28]: We should suggest uses to use `notation policy import` function to avoid manual copying or moving trust policy file to notation config directory. This command will ensure the policy file with right permissions.

File: content/en/docs/concepts/secure-deployment.md
Comment (yizha1) [inline content/en/docs/concepts/secure-deployment.md:34]: We should suggest users using `notation cert add` to add certificate to trust store and avoiding manually copying or moving. This command will ensure the certificate with right permssions.

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: thanks @yizha1 

i incorporated your feedback

File: content/en/docs/concepts/secure-deployment.md
Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md]: thanks @yizha1

i incorporated your feedback

**Review by FeynmanZhou — APPROVED** [review]
LGTM

**Review by priteshbandi — APPROVED** [review]
LGTM

File: content/en/docs/concepts/secure-deployment.md
Comment (shizhMSFT) [inline content/en/docs/concepts/secure-deployment.md:24]: I just noticed an issue that we only covered the scenarios on Linux. How about Windows and macOS?

Comment (FeynmanZhou) [reply content/en/docs/concepts/secure-deployment.md:24]: Good catch. Windows and macOS deployment with security hardening also need to be added here.

Comment (zr-msft) [reply content/en/docs/concepts/secure-deployment.md:24]: I agree, happy to do a fast follow up PR. I'm also assuming the advice for Linux will be close, if not the same for macOS

**Review by toddysm — APPROVED** [review]

**Review by iamsamirzon — APPROVED** [review]

**Pull request merged**
